### PR TITLE
Add support for deserializing list-encoded JSON structs [#6558]

### LINKED ISSUE #6558: Allow JSON deserialization of StructArray from JSON List
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently, to deserialize a StructArray from JSON, you need to use a JSON Object.  E.g., deserializing a Struct<a: i32, b: string> would need something like `{a: 1, b: "c"}`.  This is also true of top-level RecordBatches.  Some services, such as Presto and Trino, serialize ROW fields as lists.  The example above would be serialized as `[1, "c"]`.  If you already know the schema, this is a more compact representation that reduces the data on the wire.

I would like the ability for arrow_json to deserialize these list-encoded structs and record batches, perhaps under an option flag.

**Describe the solution you'd like**

When a StructArrayDecoder encounters a `[` (around [here](https://github.com/apache/arrow-rs/blob/master/arrow-json/src/reader/struct_array.rs#L75)), it switches to a parsing mode that does not look for field names, and requires a closing `]` for completion.  It should return a parsing error if either the number of entries of the list is not the same as the number of fields in the struct, or if any of the sub-parsers encounter the wrong type.  This requires the fields of the struct to be in the same order as the JSON List, while the current object parsing can shuffle the fields if they appear in a different order.

**Describe alternatives you've considered**

I currently parse the results with serde_json, then recursively run down the JSON to convert Lists to Objects using the schema.  Then I re-serialize the top-level JSON, then read it using arrow_json.  This is not very efficient.

Alternatively, I could reproduce a less-good copy of the version of arrow_json that deserialized using serde_json.  Either making serde_json decoders directly, or taking a serde_json::Value and populate the ArrayBuilders myself.  This is a lot of code duplication.

**Additional context**

I don't personally need the ability to serialize a StructArray/RecordBatch into a List, although that would seem symmetrical.

I am happy to make an RFC PR implementing this functionality.

### PR DESCRIPTION
**Pull request by jagill**
# Which issue does this PR close?

Closes #6558.

# Rationale for this change
 
Currently, a StructArray can only be deserialized from a JSON object (e.g. `{a: 1, b: "c"}`), but some services (e.g. Presto and Trino) encode ROW types as JSON lists (e.g. `[1, "c"]`) because this is more compact, and the schema is known.
Arrow-json cannot currently deserialize these.

# What changes are included in this PR?

This PR adds the ability to parse JSON lists into StructArrays, if the StructParseMode is set to ListOnly.  In ListOnly mode, object-encoded structs raise an error.  Setting to ObjectOnly (the default) has the original parsing behavior.

# Are there any user-facing changes?

Users may set the `StructParsingMode` enum to `ListOnly` to parse list-style structs.  The associated enum,
variants, and method have been documented.  I'm happy to update any other documentation.

# Discussion topics

1. I've made a JsonParseMode struct instead of a bool flag for two reasons.  One is that it's self-descriptive (what would `true` be?), and the other is that it allows a future Mixed mode that could deserialize either.  The latter isn't currently requested by anyone.
2. I kept the error messages as similar to the old messages as possible. I considered having more specific error messages (like "Encountered a '[' when parsing a Struct, but the StructParseMode is ObjectOnly" or similar), but wanted to hear opinions before I went that route.
3. I'm not attached to any name/code-style/etc, so happy to modify to fit local conventions.
4. One requirement was that benchmarks do not regress.  My running of benchmarks have been inconclusive (see https://gist.github.com/jagill/6749248171a1f12fb7c653ff70c5ed42).  There are often small regressions or improvements in the single-digit % range whenever I switch between master and this PR.  I suspect they are statistical but I wanted to note these.

### CODE REVIEW

**Comment by jagill** [comment]
Is there any changes I could make to this PR to make it easier to review?

File: arrow-json/src/reader/mod.rs [hunk 1]
```diff
@@ -170,12 +170,30 @@ mod struct_array;
 mod tape;
 mod timestamp_array;
 
+/// Specifies what is considered valid JSON when parsing StructArrays.
```
Comment (alamb) [inline arrow-json/src/reader/mod.rs:173]: Do you have a link to documentation / prior art for this type of JSON encoding of structs? Maybe to trino / presto docs that describe it in more full detail that we can add to this doc?

**Review by alamb — COMMENTED** [review]
Thank you for this contribution @jagill  and i very much apologize for the delay in reviewing -- we are always looking for more help:
- https://github.com/apache/arrow-rs/issues/6418

TLDR thoughts:
1. Are you planning / willing to implement serializing support too (aka support in json_writer)? That would then allow us to verify that data can be round-tripped
3. I believe this PR slows down performance (notes below)

# API 

> I've made a JsonParseMode struct instead of a bool flag for two reasons. One is that it's self-descriptive (what would true be?), and the other is that it allows a future Mixed mode that could deserialize either. The latter isn't currently requested by anyone.

I agree that a struct is nice for the reasons you mention

# Errors


> I kept the error messages as similar to the old messages as possible. I considered having more specific error messages (like "Encountered a '[' when parsing a Struct, but the StructParseMode is ObjectOnly" or similar), but wanted to hear opinions before I went that route.

I think improved error messages are always better, but we can do them as a follow on PR

> I'm not attached to any name/code-style/etc, so happy to modify to fit local conventions.

💯 

# Performance

> > One requirement was that benchmarks do not regress. My running of benchmarks have been inconclusive (see https://gist.github.com/jagill/6749248171a1f12fb7c653ff70c5ed42). There are often small regressions or improvements in the single-digit % range whenever I switch between master and this PR. I suspect they are statistical but I wanted to note these.

I am not sure what benchmarks you ran -- I think the most relevant ones are in `arrow/src/json_reader.rs`

I ran them like `cargo bench --bench json_reader`

Here are three back to back runs I did on my test bench:

```text
++ critcmp master json-struct-from-list
group                    json-struct-from-list                  master
-----                    ---------------------                  ------
large_bench_primitive    1.17      4.2±0.05ms        ? ?/sec    1.00      3.6±0.01ms        ? ?/sec
small_bench_list         1.03     14.5±0.05µs        ? ?/sec    1.00     14.1±0.17µs        ? ?/sec
small_bench_primitive    1.00      7.2±0.04µs        ? ?/sec    1.00      7.2±0.02µs        ? ?/sec
```

```
group                    json-struct-from-list                  master
-----                    ---------------------                  ------
large_bench_primitive    1.16      4.2±0.03ms        ? ?/sec    1.00      3.6±0.01ms        ? ?/sec
small_bench_list         1.03     14.5±0.05µs        ? ?/sec    1.00     14.0±0.13µs        ? ?/sec
small_bench_primitive    1.00      7.2±0.03µs        ? ?/sec    1.01      7.2±0.03µs        ? ?/sec
```

```
++ critcmp master json-struct-from-list
group                    json-struct-from-list                  master
-----                    ---------------------                  ------
large_bench_primitive    1.15      4.2±0.01ms        ? ?/sec    1.00      3.7±0.01ms        ? ?/sec
small_bench_list         1.00     14.4±0.10µs        ? ?/sec    1.00     14.4±0.08µs        ? ?/sec
small_bench_primitive    1.00      7.1±0.03µs        ? ?/sec    1.06      7.6±0.05µs        ? ?/sec
++ popd
~/datafusion-benchmarking
```

It shows a pretty consistent slowdown in large_bench_primitive --  I left some suggestions on how to improve this


Again, sorry for the delay in reviewing

File: arrow-json/src/reader/mod.rs (hunk 1 above)
Comment (jagill) [reply arrow-json/src/reader/mod.rs:173]: Sadly, the documentation of the format of the returned data is a little lacking.  They describe that the top-level row is a list with one entry per column (https://prestodb.io/docs/current/develop/client-protocol.html#important-queryresults-attributes, https://trino.io/docs/current/develop/client-protocol.html#important-queryresults-attributes), but do not specify how any structural elements are serialized (or non-structural things like Datetimes etc).  If it helps, I could add this to the Presto documentation.

**Comment by jagill** [comment]
Ok, I've hoisted the conditional outside the loop.  It also had the effect (as you predicted) that the gnarly match statement for the delimiters is a bit easier to understand.

> 1. Are you planning / willing to implement serializing support too (aka support in json_writer)? That would then allow us to verify that data can be round-tripped

I was not planning to, because there was no use-case requiring it.  But I'm certainly willing to, both for symmetry and for round-tripping.  Is that something you'd want in this PR, or a follow-up?

> 2. I believe this PR slows down performance (notes below)

When I check, I'm consistently seeing no difference between the main branch and either of the variants:

```
❯ critcmp main inner-conditional
group                    inner-conditional                      main
-----                    -----------------                      ----
large_bench_primitive    1.00      2.7±0.01ms        ? ?/sec    1.00      2.7±0.03ms        ? ?/sec
small_bench_list         1.01      8.4±0.09µs        ? ?/sec    1.00      8.2±0.06µs        ? ?/sec
small_bench_primitive    1.01      4.6±0.03µs        ? ?/sec    1.00      4.6±0.04µs        ? ?/sec

❯ critcmp main outer-conditional
group                    outer-conditional                      main
-----                    -----------------                      ----
large_bench_primitive    1.00      2.7±0.03ms        ? ?/sec    1.00      2.7±0.03ms        ? ?/sec
small_bench_list         1.00      8.3±0.11µs        ? ?/sec    1.00      8.2±0.06µs        ? ?/sec
small_bench_primitive    1.00      4.6±0.05µs        ? ?/sec    1.00      4.6±0.04µs        ? ?/sec

❯ critcmp inner-conditional outer-conditional
group                    outer-conditional                      inner-conditional
-----                    -----------------                      -----------------
large_bench_primitive    1.00      2.7±0.03ms        ? ?/sec    1.00      2.7±0.01ms        ? ?/sec
small_bench_list         1.00      8.3±0.11µs        ? ?/sec    1.01      8.4±0.09µs        ? ?/sec
small_bench_primitive    1.00      4.6±0.05µs        ? ?/sec    1.01      4.6±0.03µs        ? ?/sec
```

(btw, I didn't know about critcmp; thanks for the pointer!)

However, I'm not running in an isolated environment, so the differences I saw in the past seemed to be related to other processes.  I'm running on a M2 Mac and haven't tried on with x86 or other architectures, and haven't been able to isolate from the helpful random processes that Macos spins up, so I'll trust your benchmarks.

**Comment by jagill** [comment]
@alamb: With this update I've
1. included the hoisting of the conditional above the loop, which I missed last time due to atrophied `git add` muscles,
2. renamed and extracted StructParseMode to a top level StructMode in arrow_json, since I've...
3. enabled writing structs to JSON lists in the writer,
4. included some more examples in the `with_struct_mode` methods of ReaderBuilder and WriterBuilder, along with links to the scanty Presto/Trino documentation.

One thing that makes me sad is that the Writer has an `explicit_nulls` field which is not orthogonal to StructMode. In particular, writing to a JSON List _must_ include nulls, so `explicit_nulls == false` and `StructMode::ListOnly` is not a consistent state.  I saw three main approaches here:
1. Split StructMode::ObjectOnly to ObjectOnlyWithImplictNulls and ObjectOnlyWithExplicitNulls (but with better names).  This means only valid states will exist, and the Reader can choose between accepting implicit nulls or not.  But this is a breaking change.
2. Error/panic in WriterBuilder if `explicit_nulls == false` and `struct_mode == ListOnly`.  This prevents users from using a non-consistent config, but adds failure paths.
3. Only skip nulls if `explicit_nulls == false` and `struct_mode == ObjectOnly`, and add that to the documentation for `with_explicit_nulls` and `with_struct_mode`.

I chose 3 but in theory I'm open to any of these (or other suggestions).  In practice, I think that 1 would involve too much yak shaving, and we can always do the breaking change combining the two options later.

I'm also happy to add more documentation for rational or examples, with the following caveats:
1. I wanted to add a doctest but constructing an expected StructArray was so verbose it obfuscated the purpose of the doctest.  Maybe there's a better way than I know?
2. Getting the doc update PR accepted/merged into Presto and Trino will take an unknown amount of time. Maybe a week, maybe 3 months.  I'm reticent to block this update for that.

**Comment by alamb** [comment]
Awesoe -- thank you @jagill  - I will try and review this PR later today or tomorrow

File: arrow-json/src/lib.rs [hunk 2]
```diff
@@ -74,6 +74,23 @@ pub use self::writer::{ArrayWriter, LineDelimitedWriter, Writer, WriterBuilder};
...
 use serde_json::{Number, Value};
 
+/// Specifies what is considered valid JSON when parsing StructArrays.
+///
+/// If a struct with fields `("a", Int32)` and `("b", Utf8)`, it could be represented as
+/// a JSON object (`{"a": 1, "b": "c"}`) or a JSON list (`[1, "c"]`).  This enum controls
+/// which form(s) the Reader will accept.
+///
+/// For objects, the order of the key does not matter.
+/// For lists, the entries must be the same number and in the same order as the struct fields.
```
Comment (Jefffrey) [inline arrow-json/src/lib.rs:84]: Current doc here seems specific to reading/decoding; should also mention write since same enum is used as write option.

Could also put extra information such as justification for ListOnly (i.e. condensed version assuming schema is known separately, as you've explained in PR)

File: arrow-json/src/reader/struct_array.rs [hunk 3]
```diff
@@ -70,43 +74,81 @@ impl ArrayDecoder for StructArrayDecoder {
...
+                                )));
+                            }
                         }
                     }
+                    // Advance to next field
+                    cur_idx = tape.next(cur_idx + 1, "field value")?;
                 }
+            }
+        } else {
+            for (row, p) in pos.iter().enumerate() {
```
[the remark this answers is not in the document]
Comment (jagill) [reply arrow-json/src/reader/struct_array.rs:117]: Yeah, strict mode is implied for Lists, since you can't skip a field and know what field was skipped (since you have no field names).  I've updated the documentation in `with_strict_mode` to make this explicit.  It just occurs to me that `strict_mode` and `struct_mode` are very similar names.

**Comment by jagill** [comment]
@Jefffrey I addressed your comments; thanks for the feedback!

File: arrow-json/src/lib.rs [hunk 4]
```diff
@@ -74,6 +74,34 @@ pub use self::writer::{ArrayWriter, LineDelimitedWriter, Writer, WriterBuilder};
...
+/// rows formatted similarly.
+///
+/// The list encoding is more compact if the schema is known, and is used by
+/// tools such as
+/// [Presto](https://prestodb.io/docs/current/develop/client-protocol.html#important-queryresults-attributes)
+/// and [Trino](https://trino.io/docs/current/develop/client-protocol.html#important-queryresults-attributes).
+///
+/// When reading objects, the order of the key does not matter. When reading
+/// lists, the entries must be the same number and in the same order as the
+/// struct fields. Map columns are not affected by this option.
```
Comment (Jefffrey) [inline arrow-json/src/lib.rs:95]:
```suggestion
/// Specifies what is considered valid JSON when reading or writing
/// RecordBatches or StructArrays.
///
/// This enum controls which form(s) the Reader will accept and which form the
/// Writer will produce. For example, if the RecordBatch Schema is
/// `[("a", Int32), ("r", Struct([("b", Boolean), ("c", Utf8)]))]`
/// then a Reader with [`StructMode::ObjectOnly`] would read rows of the form
/// `{"a": 1, "r": {"b": true, "c": "cat"}}` while with ['StructMode::ListOnly']
/// would read rows of the form `[1, [true, "cat"]]`. A Writer would produce
/// rows formatted similarly.
///
/// The list encoding is more compact if the schema is known, and is used by
/// tools such as [Presto] and [Trino].
///
/// When reading objects, the order of the key does not matter. When reading
/// lists, the entries must be the same number and in the same order as the
/// struct fields. Map columns are not affected by this option.
///
/// [Presto]: https://prestodb.io/docs/current/develop/client-protocol.html#important-queryresults-attributes
/// [Trino]: https://trino.io/docs/current/develop/client-protocol.html#important-queryresults-attributes
```

File: arrow-json/src/reader/mod.rs [hunk 5]
```diff
@@ -253,15 +257,28 @@ impl ReaderBuilder {
         }
     }
 
-    /// Sets if the decoder should return an error if it encounters a column not present
-    /// in `schema`
+    /// Sets if the decoder should return an error if it encounters a column not
+    /// present in `schema`. If `struct_mode` is `ListOnly`, then it is required
+    /// for all fields of the struct to be in the list, regardless of the value
+    /// of `strict_mode`. Without field names, there is no way to determine
+    /// which field is missing.
```
Comment (Jefffrey) [inline arrow-json/src/reader/mod.rs:264]: Maybe explicitly state that this value essentially becomes `true` if `struct_mode` is set. Can have that first, then the explanation later (so readers can quickly determine the effect without needing the explanation).

File: arrow-json/src/reader/mod.rs [hunk 6]
```diff
@@ -2343,4 +2374,219 @@ mod tests {
             .unwrap()
         );
     }
+
+    #[test]
+    fn test_struct_decoding_list_length() {
```
[the remark this answers is not in the document]
Comment (Jefffrey) [reply arrow-json/src/reader/mod.rs:2372]: Was this actioned? I'm having a bit of trouble locating it in the test suite if so :sweat_smile:

File: arrow-json/src/reader/struct_array.rs [hunk 7]
```diff
@@ -70,43 +74,84 @@ impl ArrayDecoder for StructArrayDecoder {
...
-            let end_idx = match (tape.get(*p), nulls.as_mut()) {
-                (TapeElement::StartObject(end_idx), None) => end_idx,
-                (TapeElement::StartObject(end_idx), Some(nulls)) => {
-                    nulls.append(true);
-                    end_idx
-                }
-                (TapeElement::Null, Some(nulls)) => {
-                    nulls.append(false);
-                    continue;
+        match self.struct_mode {
```
Comment (Jefffrey) [inline arrow-json/src/reader/struct_array.rs:77]: Maybe add a note to explain why this duplication occurs, e.g.

```rust
// We avoid having the match on self.struct_mode inside the hot loop for performance
```

And then could add a TODO to investigate how to reduce the duplication if possible.

**Review by Jefffrey — COMMENTED** [review]
Sorry for the delay.

Generally looks good to me; my main concern is how introducing this new feature couples other builder options together (i.e. struct_mode with explicit_nulls and strict_mode). I guess having this documented is the best approach for now considering we don't want to break API compatibility until next major release.

File: arrow-json/src/reader/struct_array.rs (hunk 7 above)
Comment (jagill) [reply arrow-json/src/reader/struct_array.rs:79]: Re: wrong type decoding: I wrote them but left them sloppily in `test_struct_decoding_list_length`.  I've extracted them and extracted most of the logic of the parsing closure into a function.

Re: option overlapping: Yeah, without a breaking API change, I see three options:
1. Describe that strict_mode and explicit_nulls are effectively determined by struct_mode in documentation. (As is)
2. Document that the "wrong" settings for strict_mode/explicit_nulls and struct_mode is not allowed; panic in that case.
3. Explicitly set strict_mode/explicit_nulls when struct_mode is set, and checkl when struct_mode/explicit_nulls is set.  And document.

I think 3 is weird and probably leaves a footgun for later developers.  I don't have a strong opinion between either 1 or 2.

File: arrow-json/src/reader/mod.rs [hunk 8]
```diff
@@ -2343,4 +2367,294 @@ mod tests {
...
+            true,
+        );
+        let builder = if as_field {
+            ReaderBuilder::new_with_field(field.clone())
+        } else {
+            ReaderBuilder::new(Arc::new(Schema::new(vec![field].clone())))
+        };
+
+
+        */
```
Comment (Jefffrey) [inline arrow-json/src/reader/mod.rs:2409]: I think this is some accidentally leftover code?

**Review by Jefffrey — APPROVED** [review]
Generally looks good to me. We'll see if can get another opinion on the API design (possibly @alamb or @tustvold?)

> Re: option overlapping: Yeah, without a breaking API change, I see three options:
>
> 1. Describe that strict_mode and explicit_nulls are effectively determined by struct_mode in documentation. (As is)
> 2. Document that the "wrong" settings for strict_mode/explicit_nulls and struct_mode is not allowed; panic in that case.
> 3. Explicitly set strict_mode/explicit_nulls when struct_mode is set, and checkl when struct_mode/explicit_nulls is set. And document.
>
> I think 3 is weird and probably leaves a footgun for later developers. I don't have a strong opinion between either 1 or 2.

I also lean towards option 1 (current state of the PR); for future we could consider patterns like type-state builders when not constrained by not having breaking changes to the API.

File: arrow-json/src/reader/mod.rs (hunk 8 above)
Comment (jagill) [reply arrow-json/src/reader/mod.rs]: Embarrassing.  Removed.

**Comment by Jefffrey** [comment]
Thanks @jagill & @alamb

**Comment by Jefffrey** [comment]
> Generally looks good to me. We'll see if can get another opinion on the API design (possibly @alamb or @tustvold?)
> 
> > Re: option overlapping: Yeah, without a breaking API change, I see three options:
> > 
> > 1. Describe that strict_mode and explicit_nulls are effectively determined by struct_mode in documentation. (As is)
> > 2. Document that the "wrong" settings for strict_mode/explicit_nulls and struct_mode is not allowed; panic in that case.
> > 3. Explicitly set strict_mode/explicit_nulls when struct_mode is set, and checkl when struct_mode/explicit_nulls is set. And document.
> > 
> > I think 3 is weird and probably leaves a footgun for later developers. I don't have a strong opinion between either 1 or 2.
> 
> I also lean towards option 1 (current state of the PR); for future we could consider patterns like type-state builders when not constrained by not having breaking changes to the API.

Raised #7025 for future reference
